jumanji/environments/logic/graph_coloring/env.py

```diff
@@ -0,0 +1,264 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Sequence, Tuple
+
+import chex
+import jax
+import jax.numpy as jnp
+import matplotlib.animation as animation
+from jax import lax
+
+from jumanji import specs
+from jumanji.env import Environment
+from jumanji.environments.logic.graph_coloring.generator import (
+    Generator,
+    RandomGenerator,
+)
+from jumanji.environments.logic.graph_coloring.types import Observation, State
+from jumanji.environments.logic.graph_coloring.viewer import GraphColoringViewer
+from jumanji.types import TimeStep, restart, termination, transition
+
+
+class GraphColoring(Environment[State]):
+    """A JAX implementation of the Graph Coloring problem.
+
+    - observation: Observation
+        - current_node_index: jax array (int) with shape (), index of the current node
+        - adj_matrix: jax array (bool) of shape (num_nodes, num_nodes):
+            adjacency matrix of the graph
+        - colors: jax array (int32) of shape (num_nodes,):
+            each item in the array is a node in the graph and it's value denotes it's color.
+            (-1 if not assigned, otherwise 0 to num_nodes-1)
+        - action_mask: jax array (bool) of shape (num_nodes,):
+            valid colors (actions) are identified with `True`, invalid ones with `False`.
+
+    - action: int, the color to be assigned to the current node (0 to num_nodes - 1)
+
+    - reward: float
+
+    - episode termination:
+        - if all nodes have been assigned a color or if invalid action is taken, it terminates.
+
+    - state: State
+        - current_node_index: jax array (int) with shape (), index of the current node
+        - colors: jax array (int32) of shape (num_nodes,):
+            color assigned to each node, -1 if not assigned
+        - adj_matrix: jax array (bool) of shape (num_nodes, num_nodes):
+            adjacency matrix of the graph
+    """
+
+    def __init__(
+        self,
+        generator: Optional[Generator] = None,
+    ):
+        """Instantiate a `GraphColoring` environment.
+
+        Args:
+            generator: callable to instantiate environment instances.
+                Defaults to `RandomGenerator` with `num_nodes = 100`,
+                and `percent_connected = 0.8` parameters.
+        """
+        self.generator = generator or RandomGenerator(
+            num_nodes=100,
+            percent_connected=0.8,
+        )
+        num_nodes, percent_connected = self.generator.specs()
+        # Create viewer used for rendering
+        self._env_viewer = GraphColoringViewer(
+            num_nodes=num_nodes, name="Graph Coloring"
```

Comment:
From the rest of the code it seems that the `num_nodes` is assumed to be constant, it is totally fine, but I still think the viewer should not depend on it. in what @clement-bonnet suggested (possibility to have the viewer as an argument and not create it here), we wouldn't know in advance the num_nodes

Reply:
Agree with @coyettev that the viewer should not have to know `num_nodes` in advance. As for the environment, it does for now assume a constant `num_nodes`, indeed. This could be augmented by having a dynamic number of nodes and a mask that tells you which nodes exist. Yet, it is not foreseen for this version. If one wants to work with different number of nodes, one can instantiate multiple environments (cf my work on curriculum learning).

Reply:
The Viewer class is implemented using Fruchterman-Reingold force-directed algorithm to get the spring layout which is: 
```
The algorithm computes a layout by simulating the graph as a physical system,
        where nodes are repelling each other and edges are attracting connected nodes.
        The method minimizes the energy of the system over several iterations.
```
So the viewer would like to know the number of nodes available. Of course I can take the num_nodes from the State but that would be calculating the scale and color map every time render or animate are called. Do we prefer to do that as the viewer has now become a parameter passed to the env?

Reply:
@clement-bonnet & @coyettev,

Any comment on this or should I resolve the issue?

Reply:
I would say that yes it could be taken from the state, as computing the scale and color map should be quick, but happy to see what @clement-bonnet thinks

Reply:
Sorry for coming late to this. If it's only for the feature map, I think it could be inferred at render time instead of at init time. So agree with @coyettev, it would be better to not have `num_nodes` a property of the viewer.